src/otel/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+    ).clone(); // we want our own memory
+
+    let proto: Protocol;
+    match String::from_utf8_unchecked(
+        slice::from_raw_parts((*protocol).start, (*protocol).len).to_vec()
+    ).to_lowercase()
+        .as_str() {
+            "http" => proto = Protocol::HttpBinary,
+            "grpc" => proto = Protocol::Grpc,
+            e => {
+                let msg_string = format!("unknown tracer type: {:#?}", e);
+                let msg = CString::from_vec_unchecked(msg_string.as_bytes().to_vec());
+                log_callback(msg.into_raw() as _);
+                return;
+            }
+        }
+
+    // make sure we are starting with a clean state
+    nxt_otel_rs_uninit();
+
+    // Create a new mpsc channel. Tokio runtime gets receiver, the send
+    // trace function gets sender.
+    let (tx, rx): (Sender<SpanMessage>, Receiver<SpanMessage>) = mpsc::channel(32);
+
+    // Store the sender so the other function can also reach it.
+    match (*nxt_otel_rs_span_tx(false)).set(tx) {
+        /* spawn a new thread with the tokio runtime and forget about it.
+         * This function will return that allows the C code to carry on
+         * doing its thing, whereas the runtime function is a long lived
+         * process that only exits when a shutdown message is sent.
+         */
+        Ok(_) => {
+            std::thread::spawn(move || nxt_otel_rs_runtime(
+                log_callback,
+                ep,
+                proto,
+                batch_size,
+                sample_fraction,
+                rx
+            ));
+        },
+        Err(e) => {
+            let msg_string = format!("couldn't initialize tracer: {:#?}", e);
+            let msg = CString::from_vec_unchecked(msg_string.as_bytes().to_vec());
+            log_callback(msg.into_raw() as _);
+        }
+    }
+}
+
+
+/* function that we wrap around Tokio's runtime code. This is long lived,
+ * which means it stops only when a shutdown signal is sent to the rx
+ * channel, or we terminate the process and leave memory all over.
+ */
+#[tokio::main]
+async unsafe fn nxt_otel_rs_runtime(
+    log_callback: unsafe extern "C" fn(*mut i8),
+    endpoint: String,
+    proto: Protocol,
+    batch_size: f64,
+    sample_fraction: f64,
+    mut rx: Receiver<SpanMessage>
+) {
+    let pipeline = opentelemetry_otlp::new_pipeline()
+        .tracing()
+        .with_trace_config(
+            Config::default()
+                .with_resource(
+                    Resource::new(vec![KeyValue::new(
+                        opentelemetry_semantic_conventions::resource::SERVICE_NAME,
+                        "NGINX Unit",
+                    )])
+                )
+                .with_sampler(Sampler::TraceIdRatioBased(sample_fraction))
+        )
+        .with_batch_config(
+            BatchConfigBuilder::default()
+                .with_max_export_batch_size(batch_size as _)
+                .with_max_queue_size(4098)
```

Comment:
4098?

Seems like a slightly strange number, maybe you meant 4096? Or is there a specific reason for 4098?

Reply:
Does this matter? Does 4096 have any more special meaning over 4098 than being an exponent of 2?

Reply:
Well, it made me question it didn't it... and I'm sure others will...

4096 is a common buffer size. It's the same size as a page of memory (usually).

Now this doesn't look like a buffer as such, but it still begs the question of why 4098?